--- net-snmp-5.7.3+dfsg/agent/mibgroup/hardware/memory/memory_linux.c.org	2020-06-23 06:35:07.985418000 -0700
+++ net-snmp-5.7.3+dfsg/agent/mibgroup/hardware/memory/memory_linux.c	2020-06-23 09:48:14.704466000 -0700
@@ -30,7 +30,7 @@
     char        *b;
     unsigned long memtotal = 0,  memfree = 0, memshared = 0,
                   buffers = 0,   cached = 0,
-                  swaptotal = 0, swapfree = 0;
+                  swaptotal = 0, swapfree = 0, memavailable = 0;
 
     netsnmp_memory_info *mem;
 
@@ -91,6 +91,13 @@
         if (first)
             snmp_log(LOG_ERR, "No MemFree line in /proc/meminfo\n");
     }
+    b = strstr(buff, "MemAvailable: ");
+    if (b) 
+        sscanf(b, "MemAvailable: %lu", &memavailable);
+    else {
+        if (first)
+            snmp_log(LOG_ERR, "No MemAvailable line in /proc/meminfo\n");
+    }
     if (0 == netsnmp_os_prematch("Linux","2.4")) {
         b = strstr(buff, "MemShared: ");
         if (b)
@@ -148,6 +155,7 @@
         mem->units = 1024;
         mem->size  = memtotal;
         mem->free  = memfree;
+        mem->avail = memavailable;
         mem->other = -1;
     }
 
--- net-snmp-5.7.3+dfsg/agent/mibgroup/ucd-snmp/memory.c.org	2020-06-23 07:58:54.038907000 -0700
+++ net-snmp-5.7.3+dfsg/agent/mibgroup/ucd-snmp/memory.c	2020-06-23 09:49:08.417480000 -0700
@@ -106,7 +106,8 @@
             mem_info = netsnmp_memory_get_byIdx( NETSNMP_MEM_TYPE_PHYSMEM, 0 );
             if (!mem_info)
                goto NOSUCH;
-            val  =  mem_info->free;     /* memfree */
+            //val  =  mem_info->free;     /* memfree */
+            val  =  mem_info->avail;     /* available mem */
             val *= (mem_info->units/1024);
             break;
         case MEMORY_STXT_TOTAL:
--- net-snmp-5.7.3+dfsg/include/net-snmp/agent/hardware/memory.h.org	2020-06-23 09:44:08.079332000 -0700
+++ net-snmp-5.7.3+dfsg/include/net-snmp/agent/hardware/memory.h	2020-06-23 09:44:51.363058000 -0700
@@ -22,6 +22,7 @@
      long size;
      long free;
      long other;
+     long avail;
 
      netsnmp_memory_info *next;
 };
