From 25d9ad8515434cc42d5bb700f1236251756a64e9 Mon Sep 17 00:00:00 2001
From: Stepan Blyshchak <stepanb@nvidia.com>
Date: Mon, 18 Oct 2021 15:31:20 +0300
Subject: [PATCH] [schema] add ACL_TABLE_TYPE table

Signed-off-by: Stepan Blyshchak <stepanb@nvidia.com>
---
 common/schema.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/common/schema.h b/common/schema.h
index 819dad0a..e606fa0f 100644
--- a/common/schema.h
+++ b/common/schema.h
@@ -56,6 +56,7 @@ namespace swss {
 #define APP_NEIGH_SUPPRESS_VLAN_TABLE_NAME   "SUPPRESS_VLAN_NEIGH_TABLE"
 #define APP_PASS_THROUGH_ROUTE_TABLE_NAME "PASS_THROUGH_ROUTE_TABLE"
 #define APP_ACL_TABLE_TABLE_NAME          "ACL_TABLE_TABLE"
+#define APP_ACL_TABLE_TYPE_TABLE_NAME     "ACL_TABLE_TYPE_TABLE"
 #define APP_ACL_RULE_TABLE_NAME           "ACL_RULE_TABLE"
 #define APP_SFLOW_TABLE_NAME              "SFLOW_TABLE"
 #define APP_SFLOW_SESSION_TABLE_NAME      "SFLOW_SESSION_TABLE"
@@ -233,6 +234,7 @@ namespace swss {
 
 #define CFG_MIRROR_SESSION_TABLE_NAME "MIRROR_SESSION"
 #define CFG_ACL_TABLE_TABLE_NAME      "ACL_TABLE"
+#define CFG_ACL_TABLE_TYPE_TABLE_NAME "ACL_TABLE_TYPE"
 #define CFG_ACL_RULE_TABLE_NAME       "ACL_RULE"
 #define CFG_PFC_WD_TABLE_NAME         "PFC_WD"
 #define CFG_FLEX_COUNTER_TABLE_NAME "FLEX_COUNTER_TABLE"
@@ -336,6 +338,7 @@ namespace swss {
 /***** STATE DATABASE *****/
 
 #define STATE_SWITCH_CAPABILITY_TABLE_NAME          "SWITCH_CAPABILITY_TABLE"
+#define STATE_ACL_STAGE_CAPABILITY_TABLE_NAME       "ACL_STAGE_CAPABILITY_TABLE"
 #define STATE_PORT_TABLE_NAME                       "PORT_TABLE"
 #define STATE_LAG_TABLE_NAME                        "LAG_TABLE"
 #define STATE_VLAN_TABLE_NAME                       "VLAN_TABLE"
