--- 202311-w6010/src/sonic-swss/orchagent/main.cpp	2024-01-04 15:58:21.433495163 +0800
+++ 202311-w6010-bak/src/sonic-swss/orchagent/main.cpp	2024-01-04 16:11:04.108122308 +0800
@@ -598,7 +598,9 @@
         delay_factor = 2;
     }
 
-    if (gMySwitchType == "voq" || gMySwitchType == "fabric" || gMySwitchType == "chassis-packet" || gMySwitchType == "dpu" || asan_enabled)
+    /* centec create switch take longtime */
+    char *platform = getenv("platform");
+    if (gMySwitchType == "voq" || gMySwitchType == "fabric" || gMySwitchType == "chassis-packet" || gMySwitchType == "dpu" || (platform && strstr(platform, CENTEC_PLATFORM_SUBSTRING)) || asan_enabled)
     {
         /* We set this long timeout in order for orchagent to wait enough time for
          * response from syncd. It is needed since switch create takes more time
@@ -614,6 +616,10 @@
         {
             attr.value.u64 = (10 * SAI_REDIS_DEFAULT_SYNC_OPERATION_RESPONSE_TIMEOUT);
         }
+	else if (platform && strstr(platform, CENTEC_PLATFORM_SUBSTRING))
+        {
+            attr.value.u64 = (10 * SAI_REDIS_DEFAULT_SYNC_OPERATION_RESPONSE_TIMEOUT);
+        }
         else
         {
             attr.value.u64 = SAI_REDIS_DEFAULT_SYNC_OPERATION_RESPONSE_TIMEOUT;
@@ -641,11 +647,16 @@
     }
     SWSS_LOG_NOTICE("Create a switch, id:%" PRIu64, gSwitchId);
 
-    if (gMySwitchType == "voq" || gMySwitchType == "fabric" || gMySwitchType == "chassis-packet" || gMySwitchType == "dpu")
+    if (gMySwitchType == "voq" || gMySwitchType == "fabric" || gMySwitchType == "chassis-packet" || gMySwitchType == "dpu" || (platform && strstr(platform, CENTEC_PLATFORM_SUBSTRING)))
     {
         /* Set syncd response timeout back to the default value */
         attr.id = SAI_REDIS_SWITCH_ATTR_SYNC_OPERATION_RESPONSE_TIMEOUT;
         attr.value.u64 = SAI_REDIS_DEFAULT_SYNC_OPERATION_RESPONSE_TIMEOUT;
+	/* centec platform set 3X default timeout*/
+        if (platform && strstr(platform, CENTEC_PLATFORM_SUBSTRING))
+        {
+            attr.value.u64 = (3 * SAI_REDIS_DEFAULT_SYNC_OPERATION_RESPONSE_TIMEOUT);
+        }
         status = sai_switch_api->set_switch_attribute(gSwitchId, &attr);
 
         if (status != SAI_STATUS_SUCCESS)
