diff -uprN sonic-sairedis-202211-bak/configure.ac sonic-sairedis-202211/configure.ac
--- sonic-sairedis-202211-bak/configure.ac	2023-04-19 15:09:07.721537685 +0800
+++ sonic-sairedis-202211/configure.ac	2023-04-19 14:28:52.227718626 +0800
@@ -188,30 +188,6 @@ AC_SUBST(CXXFLAGS_COMMON)
 AC_SUBST(SAIINC, "-I\$(top_srcdir)/SAI/inc -I\$(top_srcdir)/SAI/experimental -I\$(top_srcdir)/SAI/meta")
 
 AM_COND_IF([SYNCD], [
-AM_COND_IF([SAIVS], [], [
-SAVED_FLAGS="$CXXFLAGS"
-CXXFLAGS="-Xlinker --no-as-needed -lsai -I$srcdir/SAI/inc -I$srcdir/SAI/experimental -I$srcdir/SAI/meta"
-AC_CHECK_FUNCS(sai_query_api_version, [
-AC_MSG_CHECKING([SAI headers API version and library version check])
-AC_TRY_RUN([
-extern "C" {
-#include <sai.h>
-}
-int main() {
-    sai_api_version_t version;
-    if (SAI_STATUS_SUCCESS != sai_query_api_version(&version))
-    {
-        return 1;
-    }
-    return (version != SAI_API_VERSION);
-}],
-[AC_MSG_RESULT(ok)],
-[AC_MSG_RESULT(failed)
-AC_MSG_ERROR("SAI headers API version and library version mismatch")])])
-CXXFLAGS="$SAVED_FLAGS"
-])])
-
-AM_COND_IF([SYNCD], [
 SAVED_FLAGS="$CXXFLAGS"
 CXXFLAGS="-Xlinker --no-as-needed -lsai -I$srcdir/SAI/inc -I$srcdir/SAI/experimental -I$srcdir/SAI/meta"
 AC_CHECK_FUNCS(sai_bulk_object_clear_stats sai_bulk_object_get_stats)
Binary files sonic-sairedis-202211-bak/.git/index and sonic-sairedis-202211/.git/index differ
diff -uprN sonic-sairedis-202211-bak/.git/logs/HEAD sonic-sairedis-202211/.git/logs/HEAD
--- sonic-sairedis-202211-bak/.git/logs/HEAD	2023-04-19 15:09:08.025538553 +0800
+++ sonic-sairedis-202211/.git/logs/HEAD	2023-04-19 14:18:37.639149332 +0800
@@ -3,4 +3,3 @@
 58e7c4fa6b1916d9f26e4e5b5625879a5745127d 62aa2448328f50815336561e141d076cf03d4087 pettershao-ragilenetworks <pettershao@ragilenetworks.com> 1681867042 +0800	reset: moving to origin/202211
 62aa2448328f50815336561e141d076cf03d4087 21f1312d860eeb3fadd606bdaf1133546e38741e pettershao-ragilenetworks <pettershao@ragilenetworks.com> 1681867308 +0800	commit: update-configure-ac
 21f1312d860eeb3fadd606bdaf1133546e38741e 62aa2448328f50815336561e141d076cf03d4087 pettershao-ragilenetworks <pettershao@ragilenetworks.com> 1681885117 +0800	reset: moving to origin/202211
-62aa2448328f50815336561e141d076cf03d4087 62aa2448328f50815336561e141d076cf03d4087 pettershao-ragilenetworks <pettershao@ragilenetworks.com> 1681888148 +0800	reset: moving to origin/202211
diff -uprN sonic-sairedis-202211-bak/.git/ORIG_HEAD sonic-sairedis-202211/.git/ORIG_HEAD
--- sonic-sairedis-202211-bak/.git/ORIG_HEAD	2023-04-19 15:09:08.025538553 +0800
+++ sonic-sairedis-202211/.git/ORIG_HEAD	2023-04-19 14:18:37.639149332 +0800
@@ -1 +1 @@
-62aa2448328f50815336561e141d076cf03d4087
+21f1312d860eeb3fadd606bdaf1133546e38741e
diff -uprN sonic-sairedis-202211-bak/lib/sai_redis_interfacequery.cpp sonic-sairedis-202211/lib/sai_redis_interfacequery.cpp
--- sonic-sairedis-202211-bak/lib/sai_redis_interfacequery.cpp	2023-04-19 15:09:07.729537708 +0800
+++ sonic-sairedis-202211/lib/sai_redis_interfacequery.cpp	2023-04-19 14:31:40.648215843 +0800
@@ -242,14 +242,6 @@ sai_status_t sai_query_stats_capability(
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
 
-sai_status_t sai_query_api_version(
-        _Out_ sai_api_version_t *version)
-{
-    SWSS_LOG_ENTER();
-
-    return SAI_STATUS_NOT_IMPLEMENTED;
-}
-
 sai_status_t sai_bulk_object_get_stats(
         _In_ sai_object_id_t switch_id,
         _In_ sai_object_type_t object_type,
diff -uprN sonic-sairedis-202211-bak/syncd/VendorSai.cpp sonic-sairedis-202211/syncd/VendorSai.cpp
--- sonic-sairedis-202211-bak/syncd/VendorSai.cpp	2023-04-19 15:09:07.753537777 +0800
+++ sonic-sairedis-202211/syncd/VendorSai.cpp	2023-04-19 14:41:20.944247597 +0800
@@ -5,7 +5,6 @@
 
 #include "swss/logger.h"
 
-#include <cinttypes>
 #include <cstring>
 
 using namespace syncd;
@@ -59,27 +58,6 @@ sai_status_t VendorSai::initialize(
         return SAI_STATUS_INVALID_PARAMETER;
     }
 
-#ifdef HAVE_SAI_QUERY_API_VERSION
-    sai_api_version_t version{};
-    auto api_status = sai_query_api_version(&version);
-    if (api_status != SAI_STATUS_SUCCESS)
-    {
-        SWSS_LOG_ERROR("failed to query SAI API version");
-
-        return api_status;
-    }
-
-    SWSS_LOG_NOTICE("SAI API version: %" PRId64, version);
-
-    if (version != SAI_API_VERSION)
-    {
-        SWSS_LOG_ERROR("SAI implementation API version %" PRId64 " does not match SAI headers API version %" PRId64,
-                       version, SAI_API_VERSION);
-
-        return SAI_STATUS_FAILURE;
-    }
-#endif
-
     memcpy(&m_service_method_table, service_method_table, sizeof(m_service_method_table));
 
     auto status = sai_api_initialize(flags, service_method_table);
diff -uprN sonic-sairedis-202211-bak/vslib/sai_vs_interfacequery.cpp sonic-sairedis-202211/vslib/sai_vs_interfacequery.cpp
--- sonic-sairedis-202211-bak/vslib/sai_vs_interfacequery.cpp	2023-04-19 15:09:08.021538542 +0800
+++ sonic-sairedis-202211/vslib/sai_vs_interfacequery.cpp	2023-04-19 15:02:54.334145907 +0800
@@ -194,15 +194,6 @@ sai_status_t sai_query_stats_capability(
     return vs_sai->queryStatsCapability(switch_id, object_type, stats_capability);
 }
 
-sai_status_t sai_query_api_version(
-        _Out_ sai_api_version_t *version)
-{
-    SWSS_LOG_ENTER();
-
-    *version = SAI_API_VERSION;
-    return SAI_STATUS_SUCCESS;
-}
-
 sai_status_t sai_bulk_object_get_stats(
         _In_ sai_object_id_t switch_id,
         _In_ sai_object_type_t object_type,
