From 21f1312d860eeb3fadd606bdaf1133546e38741e Mon Sep 17 00:00:00 2001
From: pettershao-ragilenetworks <pettershao@ragilenetworks.com>
Date: Wed, 19 Apr 2023 09:21:48 +0800
Subject: [PATCH] update-configure-ac

---
 configure.ac | 17 -----------------
 1 file changed, 17 deletions(-)

diff --git a/configure.ac b/configure.ac
index e7ccb4e..979dc00 100644
--- a/configure.ac
+++ b/configure.ac
@@ -191,23 +191,6 @@ AM_COND_IF([SYNCD], [
 AM_COND_IF([SAIVS], [], [
 SAVED_FLAGS="$CXXFLAGS"
 CXXFLAGS="-Xlinker --no-as-needed -lsai -I$srcdir/SAI/inc -I$srcdir/SAI/experimental -I$srcdir/SAI/meta"
-AC_CHECK_FUNCS(sai_query_api_version, [
-AC_MSG_CHECKING([SAI headers API version and library version check])
-AC_TRY_RUN([
-extern "C" {
-#include <sai.h>
-}
-int main() {
-    sai_api_version_t version;
-    if (SAI_STATUS_SUCCESS != sai_query_api_version(&version))
-    {
-        return 1;
-    }
-    return (version != SAI_API_VERSION);
-}],
-[AC_MSG_RESULT(ok)],
-[AC_MSG_RESULT(failed)
-AC_MSG_ERROR("SAI headers API version and library version mismatch")])])
 CXXFLAGS="$SAVED_FLAGS"
 ])])
 
-- 
2.25.1

