--- sonic-buildimage-2/src/sonic-linux-kernel/patch/series	2023-02-14 13:54:19.920742540 +0800
+++ sonic-buildimage-1/src/sonic-linux-kernel/patch/series	2023-02-14 13:57:06.792842786 +0800
@@ -166,6 +166,10 @@ cisco-hwmon-pmbus_core-pec-support-check
 #
 # Marvell platform patches for 4.19
 armhf_secondary_boot_online.patch
+
+#
+# Ragile patches for kernel
+0001-mm-fix-exec-activate_mm-vs-TLB-shootdown-and-lazy-tl.patch
 #
 #
 ############################################################
