From ad4136ca0e9f37861b17e15ec69860486ac00268 Mon Sep 17 00:00:00 2001
From: liuh <liuh@microsoft.com>
Date: Fri, 1 Oct 2021 22:49:01 +0800
Subject: [PATCH] Add setting flag for authorization and accounting.

---
 support.c | 12 ++++++++++--
 support.h |  8 ++++++++
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/support.c b/support.c
index 4be5339..d065023 100644
--- a/support.c
+++ b/support.c
@@ -324,7 +324,7 @@ int _pam_parse_arg (const char *arg, char* current_secret, uint current_secret_b
 
         /* if 'secret=' was given after a 'server=' parameter, fill in the current secret */
         for(i = tac_srv_no-1; i >= 0; i--) {
-            if (tac_srv[i].key != NULL)
+            if (tac_srv[i].key[0] != 0)
                 break;
 
             /* copy secret to key */
@@ -344,7 +344,15 @@ int _pam_parse_arg (const char *arg, char* current_secret, uint current_secret_b
     } else if (!strncmp (arg, "source_ip=", strlen("source_ip="))) {
         /* source ip for the packets */
         strncpy (tac_source_ip, arg + strlen("source_ip="), sizeof(tac_source_ip));
-        set_source_ip (tac_source_ip);
+        set_source_ip (tac_source_ip);
+    } else if (!strcmp (arg, "local_accounting")) {
+        ctrl |= ACCOUNTING_FLAG_LOCAL;
+    } else if (!strcmp (arg, "tacacs_accounting")) {
+        ctrl |= ACCOUNTING_FLAG_TACACS;
+    } else if (!strcmp (arg, "local_authorization")) {
+        ctrl |= AUTHORIZATION_FLAG_LOCAL;
+    } else if (!strcmp (arg, "tacacs_authorization")) {
+        ctrl |= AUTHORIZATION_FLAG_TACACS;
     } else {
         _pam_log (LOG_WARNING, "unrecognized option: %s", arg);
     }
diff --git a/support.h b/support.h
index 6820be3..7862e58 100644
--- a/support.h
+++ b/support.h
@@ -32,6 +32,14 @@
 #define PAM_TAC_USE_FIRST_PASS 0x04
 #define PAM_TAC_TRY_FIRST_PASS 0x08
 
+/* accounting setting flag */
+#define ACCOUNTING_FLAG_LOCAL  0x10
+#define ACCOUNTING_FLAG_TACACS 0x20
+
+/* authorization setting flag */
+#define AUTHORIZATION_FLAG_LOCAL  0x40
+#define AUTHORIZATION_FLAG_TACACS 0x80
+
 typedef struct {
     struct addrinfo *addr;
     char key[256];
-- 
2.27.0.windows.1

