From c755ff147a89206c92f02d48a91d7540c6768be7 Mon Sep 17 00:00:00 2001
From: Wenda <wenni@microsoft.com>
Date: Fri, 12 Oct 2018 02:47:58 +0000
Subject: [PATCH] Patch to change dynamic alpha at run time

Signed-off-by: Wenda <wenni@microsoft.com>
---
 orchagent/bufferorch.cpp | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/orchagent/bufferorch.cpp b/orchagent/bufferorch.cpp
index ef77aa8..c399b17 100644
--- a/orchagent/bufferorch.cpp
+++ b/orchagent/bufferorch.cpp
@@ -340,6 +340,25 @@ task_process_status BufferOrch::processBufferProfile(Consumer &consumer)
         if (SAI_NULL_OBJECT_ID != sai_object)
         {
             SWSS_LOG_DEBUG("Modifying existing sai object:%lx ", sai_object);
+            // Filter those attributes that do not support runtime modification
+            auto it = attribs.begin();
+            while (it != attribs.end())
+            {
+                switch (it->id)
+                {
+                    case SAI_BUFFER_PROFILE_ATTR_SHARED_DYNAMIC_TH:
+                        ++it;
+                        break;
+                    default:
+                        it = attribs.erase(it);
+                        break;
+                }
+            }
+            if (0 == attribs.size())
+            {
+                SWSS_LOG_ERROR("Missing or invalid buffer profile attributes for runtime modification, name:%s, sai object:%lx", object_name.c_str(), sai_object);
+                return task_process_status::task_invalid_entry;
+            }
             sai_status = sai_buffer_api->set_buffer_profile_attribute(sai_object, &attribs[0]);
             if (SAI_STATUS_SUCCESS != sai_status)
             {
-- 
2.7.4

