From e213bbf6dc8f1431ebb9ab9d1d57dcb6f37b5b20 Mon Sep 17 00:00:00 2001
From: Tamer Ahmed <tamer.ahmed@microsoft.com>
Date: Wed, 8 Jan 2020 10:05:16 -0800
Subject: [PATCH 2/3] [SNMP] Disable SNMPv1

signed-off-by: Tamer Ahmed <tamer.ahmed@microsoft.com>
---
 debian/rules | 1 +
 1 file changed, 1 insertion(+)

diff --git a/debian/rules b/debian/rules
index 181ad5c..8778221 100755
--- a/debian/rules
+++ b/debian/rules
@@ -37,6 +37,7 @@ override_dh_auto_build:
 override_dh_auto_configure:
 	dh_auto_configure -- --prefix=/usr --sysconfdir=/etc --mandir=/usr/share/man \
 	  --with-persistent-directory=/var/lib/snmp \
+	  --disable-snmpv1 \
 	  --enable-ucd-snmp-compatibility \
 	  --with-cflags="$(CFLAGS) -DNETSNMP_USE_INLINE" \
 	  --with-ldflags="$(LDFLAGS)" \
-- 
2.17.1

