From 839f98d579e3a22b6ca075fd026eee55bdb2ef25 Mon Sep 17 00:00:00 2001
From: Blueve <672454911@qq.com>
Date: Mon, 26 Oct 2020 06:44:59 +0000
Subject: [PATCH] Put style as line number to ssh session environment variable

---
 session.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/session.c b/session.c
index 19f38637e..654371447 100644
--- a/session.c
+++ b/session.c
@@ -1209,6 +1209,11 @@ do_setup_env(struct ssh *ssh, Session *s, const char *shell)
 		child_set_env(&env, &envsize, "SSH_ORIGINAL_COMMAND",
 		    original_command);
 
+	/* Take advantage of authentication style field */
+	if (s->authctxt->style)
+		child_set_env(&env, &envsize, "SSH_TARGET_CONSOLE_LINE",
+		    s->authctxt->style);
+
 	if (debug_flag) {
 		/* dump the environment */
 		fprintf(stderr, "Environment:\n");
-- 
2.25.1

