diff --git a/ospfd/ospf_vty.c b/ospfd/ospf_vty.c
index 387472454..122d8525e 100755
--- a/ospfd/ospf_vty.c
+++ b/ospfd/ospf_vty.c
@@ -3927,15 +3927,15 @@ static int show_ip_ospf_interface_common(struct vty *vty, struct ospf *ospf,
 		/* Interface name is specified. */
 		ifp = if_lookup_by_name(intf_name, ospf->vrf_id);
 		if (ifp == NULL) {
-			if (use_json)
+			if (use_json) {
 				json_object_boolean_true_add(json_vrf,
 							     "noSuchIface");
-			else
+				json_object_free(json_interface);
+			} else
 				vty_out(vty, "No such interface name\n");
 		} else {
 			if (use_json) {
 				json_interface_sub = json_object_new_object();
-				json_interface = json_object_new_object();
 			}
 
 			show_ip_ospf_interface_sub(
@@ -4070,6 +4070,10 @@ static int show_ip_ospf_interface_traffic_common(
 				vty_out(vty,
 					"  OSPF not enabled on this interface %s\n",
 					ifp->name);
+				if (use_json) {
+					if (use_vrf)
+						json_object_free(json_vrf);
+				}
 				return CMD_SUCCESS;
 			}
 
@@ -7024,8 +7028,13 @@ static int show_ip_ospf_database_common(struct vty *vty, struct ospf *ospf,
 		type = OSPF_OPAQUE_AREA_LSA;
 	else if (strncmp(argv[arg_base + idx_type]->text, "opaque-as", 9) == 0)
 		type = OSPF_OPAQUE_AS_LSA;
-	else
+	else {
+		if (uj) {
+			if (use_vrf)
+				json_object_free(json_vrf);
+		}
 		return CMD_WARNING;
+	}
 
 	/* `show ip ospf database LSA'. */
 	if ((argc == arg_base + 5) || (uj && (argc == arg_base + 6)))
@@ -11028,7 +11037,12 @@ static int show_ip_ospf_route_common(struct vty *vty, struct ospf *ospf,
 	ospf_show_vrf_name(ospf, vty, json_vrf, use_vrf);
 
 	if (ospf->new_table == NULL) {
-		vty_out(vty, "No OSPF routing information exist\n");
+		if (json) {
+			if (use_vrf)
+				json_object_free(json_vrf);
+		} else {
+			vty_out(vty, "No OSPF routing information exist\n");
+		}
 		return CMD_SUCCESS;
 	}
 
