From df79586612fb0445fbdf6b191747294e5494ece2 Mon Sep 17 00:00:00 2001
From: Christian Hopps <chopps@labn.net>
Date: Thu, 26 Oct 2023 22:51:08 -0400
Subject: [PATCH 01/11] isisd: staticd: need to link directly against libyang

Signed-off-by: Christian Hopps <chopps@labn.net>
(cherry picked from commit 81d1d399521bb18f3fdd5353c9d58c4b3988f225)

diff --git a/isisd/subdir.am b/isisd/subdir.am
index dabf6a925e..10f9f5b964 100644
--- a/isisd/subdir.am
+++ b/isisd/subdir.am
@@ -92,7 +92,7 @@ ISIS_SOURCES = \
 	isisd/isis_pfpacket.c \
 	# end
 
-ISIS_LDADD_COMMON = lib/libfrr.la $(LIBCAP)
+ISIS_LDADD_COMMON = lib/libfrr.la $(LIBCAP) $(LIBYANG_LIBS)
 
 # Building isisd
 
diff --git a/staticd/subdir.am b/staticd/subdir.am
index 022428281f..07ebe3c02c 100644
--- a/staticd/subdir.am
+++ b/staticd/subdir.am
@@ -36,7 +36,7 @@ clippy_scan += \
 	# end
 
 staticd_staticd_SOURCES = staticd/static_main.c
-staticd_staticd_LDADD = staticd/libstatic.a lib/libfrr.la $(LIBCAP)
+staticd_staticd_LDADD = staticd/libstatic.a lib/libfrr.la $(LIBCAP) $(LIBYANG_LIBS)
 
 nodist_staticd_staticd_SOURCES = \
 	yang/frr-bfdd.yang.c \
-- 
2.17.1

