From 0139af1f4b07a2f2305aa93df1b43fbdfe8893a9 Mon Sep 17 00:00:00 2001
From: xuliping <xuliping@microsoft.com>
Date: Mon, 4 Mar 2024 12:09:31 +0800
Subject: [PATCH 2/2] update

---
 bgpd/bgp_updgrp_adv.c | 20 ++++++++++++--------
 1 file changed, 12 insertions(+), 8 deletions(-)

diff --git a/bgpd/bgp_updgrp_adv.c b/bgpd/bgp_updgrp_adv.c
index 17b126a8ad..22af326230 100644
--- a/bgpd/bgp_updgrp_adv.c
+++ b/bgpd/bgp_updgrp_adv.c
@@ -670,7 +670,7 @@ void subgroup_announce_table(struct update_subgroup *subgrp,
 {
 	struct bgp_dest *dest;
 	struct bgp_path_info *ri;
-	struct attr attr;
+	struct attr attr = {0};
 	struct peer *peer;
 	afi_t afi;
 	safi_t safi;
@@ -717,17 +717,20 @@ void subgroup_announce_table(struct update_subgroup *subgrp,
 						    &attr, NULL)) {
 				/* Check if route can be advertised */
 				if (advertise) {
-					if (!bgp_check_withdrawal(bgp, dest))
-						bgp_adj_out_set_subgroup(
+					if (!bgp_check_withdrawal(bgp, dest)) {
+						if (!bgp_adj_out_set_subgroup(
 							dest, subgrp, &attr,
-							ri);
-					else
+							ri))
+							bgp_attr_flush(&attr);
+					} else {
 						bgp_adj_out_unset_subgroup(
 							dest, subgrp, 1,
 							bgp_addpath_id_for_peer(
 								peer, afi,
 								safi_rib,
 								&ri->tx_addpath));
+						bgp_attr_flush(&attr);
+					}
 				}
 			} else {
 				/* If default originate is enabled for
@@ -747,6 +750,7 @@ void subgroup_announce_table(struct update_subgroup *subgrp,
 					bgp_addpath_id_for_peer(
 						peer, afi, safi_rib,
 						&ri->tx_addpath));
+				bgp_attr_flush(&attr);
 			}
 		}
 	}
@@ -957,10 +961,10 @@ void subgroup_default_originate(struct update_subgroup *subgrp, int withdraw)
 						struct attr *default_attr =
 							bgp_attr_intern(&attr);
 
-						if (!bgp_adj_out_set_subgroup(
+						bgp_adj_out_set_subgroup(
 							    dest, subgrp,
-							    default_attr, pi))
-							bgp_attr_flush(&attr);
+							    default_attr, pi);
+						bgp_attr_unintern(&default_attr);
 					} else
 						bgp_attr_flush(&attr);
 				}
-- 
2.25.1

