From d2c7ec53c6e4360183d593dd886b68bf30ec4f5d Mon Sep 17 00:00:00 2001
From: Joe LeVeque <jolevequ@microsoft.com>
Date: Tue, 19 Dec 2017 23:58:37 +0000
Subject: [PATCH 1/4] Do not discard VLAN-tagged packets

---
 common/lpf.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/common/lpf.c b/common/lpf.c
index 7889b6b..3789d54 100644
--- a/common/lpf.c
+++ b/common/lpf.c
@@ -425,8 +425,8 @@ ssize_t receive_packet (interface, buf, len, from, hfrom)
 			/* Discard packets with stripped vlan id */
 
 #ifdef VLAN_TCI_PRESENT
-			if (aux->tp_vlan_tci != 0)
-				return 0;
+			//if (aux->tp_vlan_tci != 0)
+			//	return 0;
 #endif
 
 			csum_ready = ((aux->tp_status & TP_STATUS_CSUMNOTREADY)
-- 
2.1.4

