From 6ee1b280c1bf06b5dedef33522eccfbcb70d02bd Mon Sep 17 00:00:00 2001
From: Joe LeVeque <jolevequ@microsoft.com>
Date: Sun, 5 May 2019 22:52:49 +0000
Subject: [PATCH 2/2] If using a socket per interface, send to ALL servers on
 ALL upstream interfaces

---
 relay/dhcrelay.c | 38 ++++++++++++++++++++++++++++----------
 1 file changed, 28 insertions(+), 10 deletions(-)

diff --git a/relay/dhcrelay.c b/relay/dhcrelay.c
index 292ba4f..72bbe1c 100644
--- a/relay/dhcrelay.c
+++ b/relay/dhcrelay.c
@@ -799,18 +799,36 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
 	/* Otherwise, it's a BOOTREQUEST, so forward it to all the
 	   servers. */
 	for (sp = servers; sp; sp = sp->next) {
-		if (send_packet((fallback_interface
-				 ? fallback_interface : interfaces),
-				 NULL, packet, length, ip->addresses[0],
-				 &sp->to, NULL) < 0) {
-			++client_packet_errors;
+		if (fallback_interface) {
+			if (send_packet(fallback_interface, NULL, packet,
+					length, ip->addresses[0],&sp->to, NULL) < 0) {
+				++client_packet_errors;
+			} else {
+				log_debug("Forwarded BOOTREQUEST for %s to %s on fallback interface",
+				       print_hw_addr(packet->htype, packet->hlen,
+						      packet->chaddr),
+				       inet_ntoa(sp->to.sin_addr));
+
+				++client_packets_relayed;
+			}
 		} else {
-			log_debug("Forwarded BOOTREQUEST for %s to %s",
-			       print_hw_addr(packet->htype, packet->hlen,
-					      packet->chaddr),
-			       inet_ntoa(sp->to.sin_addr));
+			for (out = interfaces; out; out = out->next) {
+				// Only relay BOOTREQUEST on upstream interfaces
+				if (!(out->flags & INTERFACE_UPSTREAM))
+					continue;
+
+				if (send_packet(out, NULL, packet,
+						length, ip->addresses[0],&sp->to, NULL) < 0) {
+					++client_packet_errors;
+				} else {
+					log_debug("Forwarded BOOTREQUEST for %s to %s on interface %s",
+					       print_hw_addr(packet->htype, packet->hlen,
+							      packet->chaddr),
+					       inet_ntoa(sp->to.sin_addr), out->name);
 
-			++client_packets_relayed;
+					++client_packets_relayed;
+				}
+			}
 		}
 	}
 				 
-- 
2.17.1

