From 7d0a63278aa5eedadd48db3937fa0139f3411669 Mon Sep 17 00:00:00 2001
From: Vaibhav Hemant Dixit <vadixit@microsoft.com>
Date: Fri, 23 Jul 2021 16:27:56 -0700
Subject: [PATCH] [fast-reboot][patch] Restore SLB neighbors state after
 fast-reboot failure

---
 scripts/fast-reboot | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

Index: sonic-utilities/scripts/fast-reboot
===================================================================
--- sonic-utilities.orig/scripts/fast-reboot
+++ sonic-utilities/scripts/fast-reboot
@@ -36,6 +36,7 @@ EXIT_ORCHAGENT_SHUTDOWN=10
 EXIT_FILTER_FDB_ENTRIES_FAILURE=13
 EXIT_COUNTERPOLL_DELAY_FAILURE=14
 EXIT_DB_INTEGRITY_FAILURE=15
+EXIT_HANDLE_SLB_NEIGHBORS=200
 EXIT_NO_CONTROL_PLANE_ASSISTANT=20
 EXIT_SONIC_INSTALLER_VERIFY_REBOOT=21
 
@@ -122,6 +123,8 @@ function clear_fast_boot()
     common_clear
 
     sonic-db-cli STATE_DB DEL "FAST_REBOOT|system" &>/dev/null || /bin/true
+
+    handle_slb_neighbors restore
 }
 
 function clear_warm_boot()
@@ -420,6 +423,32 @@ function check_warm_restart_in_progress(
     done
 }
 
+function handle_slb_neighbors()
+{
+    # Revert all SLB neighbors if they're presented
+    if [[ "$REBOOT_TYPE" = "fast-reboot" ]]; then
+        debug "Stopping all SLB neighbors if they are presented"
+        PASSIVE_BGP_NEIGHBORS=$(sonic-cfggen -d -v "BGP_PEER_RANGE | list")
+        case "$PASSIVE_BGP_NEIGHBORS" in
+            *BGPSLBPassive*)
+            RUNNING_CONFIG=$(vtysh -c "show running")
+            if grep -q "neighbor BGPSLBPassive peer-group" <<< "$RUNNING_CONFIG"; then
+                ASN=$(sonic-cfggen -d -v "DEVICE_METADATA['localhost']['bgp_asn']")
+                case "$1" in
+                        *shutdown) CMD="neighbor BGPSLBPassive shutdown";;
+                        *restore) CMD="no neighbor BGPSLBPassive shutdown";;
+                        *) debug "Wrong/missing argument to handle_slb_neighbors"; exit "${EXIT_HANDLE_SLB_NEIGHBORS}";;
+                esac
+                vtysh -c "configure terminal" -c "router bgp ${ASN}" -c "${CMD}"
+                sleep 30 # wait for 30 seconds - BGP RouteAdv default timer
+            fi
+            ;;
+        *)
+            ;;
+        esac
+    fi
+}
+
 # main starts here
 parseOptions $@
 
@@ -460,6 +489,8 @@ case "$REBOOT_TYPE" in
         ;;
 esac
 
+handle_slb_neighbors shutdown
+
 save_counters_folder
 
 unload_kernel
