diff --git a/debian/kdump-config b/debian/kdump-config
index aa31e41..3a88d21 100755
--- a/debian/kdump-config
+++ b/debian/kdump-config
@@ -43,7 +43,7 @@ KVER=`uname -r`
 ARCH=`uname -m`
 
 # Set up defaults
-KDUMP_SYSCTL=${KDUMP_SYSCTL:="kernel.panic_on_oops=1"}
+KDUMP_SYSCTL=${KDUMP_SYSCTL:="kernel.panic_on_oops=1 kernel.panic_on_unrecovered_nmi=1 kernel.panic_on_io_nmi=1 kernel.panic_on_stackoverflow=1 kernel.hung_task_panic=1 kernel.softlockup_panic=1 kernel.unknown_nmi_panic=1 vm.panic_on_oom=1"}
 KDUMP_COREDIR=${KDUMP_COREDIR:=/var/crash}
 KDUMP_DUMP_DMESG=${KDUMP_DUMP_DMESG:=1}
 KDUMP_DIR="/var/lib/kdump"
