From 994a54bbe8a4b2514da99221d80c8f7d5d51f12b Mon Sep 17 00:00:00 2001

From: Jiahua Wang <jiahwang@cisco.com>

Subject: [PATCH 2/2] Increase ARCH_NR_GPIOs to 1024

Signed-off-by: Jiahua Wang <jiahwang@cisco.com>
---
 include/asm-generic/gpio.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/asm-generic/gpio.h b/include/asm-generic/gpio.h
index 8ca627d..43ad6d7 100644
--- a/include/asm-generic/gpio.h
+++ b/include/asm-generic/gpio.h
@@ -29,7 +29,7 @@
 #if defined(CONFIG_ARCH_NR_GPIO) && CONFIG_ARCH_NR_GPIO > 0
 #define ARCH_NR_GPIOS CONFIG_ARCH_NR_GPIO
 #else
-#define ARCH_NR_GPIOS		512
+#define ARCH_NR_GPIOS		1024
 #endif
 #endif
 
