Description: fix PHPUnit 6+ extend format
 Subclassed TestCase.
Author: Laszlo Boszormenyi (GCS) <gcs@debian.org>
Last-Update: 2018-09-20

---

--- thrift-0.11.0.orig/lib/php/test/Test/Thrift/Protocol/TestTJSONProtocol.php
+++ thrift-0.11.0/lib/php/test/Test/Thrift/Protocol/TestTJSONProtocol.php
@@ -46,7 +46,7 @@ $loader->register();
  *   --out ./packages ../../../test/ThriftTest.thrift
  */
 
-class TestTJSONProtocol extends \PHPUnit_Framework_TestCase
+class TestTJSONProtocol extends \PHPUnit\Framework\TestCase
 {
   private $transport;
   private $protocol;
--- thrift-0.11.0.orig/lib/php/test/Test/Thrift/Protocol/TestBinarySerializer.php
+++ thrift-0.11.0/lib/php/test/Test/Thrift/Protocol/TestBinarySerializer.php
@@ -42,7 +42,7 @@ $loader->register();
  *   --out ./packages ../../../test/ThriftTest.thrift
  */
 
-class TestBinarySerializer extends \PHPUnit_Framework_TestCase
+class TestBinarySerializer extends \PHPUnit\Framework\TestCase
 {
 
   public function setUp()
--- thrift-0.11.0.orig/lib/php/test/Test/Thrift/JsonSerialize/JsonSerializeTest.php
+++ thrift-0.11.0/lib/php/test/Test/Thrift/JsonSerialize/JsonSerializeTest.php
@@ -31,7 +31,7 @@ $loader->registerNamespace('Test', __DIR
 $loader->registerDefinition('ThriftTest', __DIR__ . '/../../../packages/phpjs');
 $loader->register();
 
-class JsonSerializeTest extends \PHPUnit_Framework_TestCase
+class JsonSerializeTest extends \PHPUnit\Framework\TestCase
 {
   protected function setUp() {
     if (version_compare(phpversion(), '5.4', '<')) {
--- thrift-0.11.0.orig/lib/php/test/Test/Thrift/Protocol/TestTSimpleJSONProtocol.php
+++ thrift-0.11.0/lib/php/test/Test/Thrift/Protocol/TestTSimpleJSONProtocol.php
@@ -46,7 +46,7 @@ $loader->register();
  *   --out ./packages ../../../test/ThriftTest.thrift
  */
 
-class TestTSimpleJSONProtocol extends \PHPUnit_Framework_TestCase
+class TestTSimpleJSONProtocol extends \PHPUnit\Framework\TestCase
 {
   private $transport;
   private $protocol;
